tensorstore/kvstore/s3_sdk/s3_context.cc

```diff
@@ -0,0 +1,294 @@
+    // TODO: This is impractical for large data and
+    // should be mitigated by an Aws::IOStream backed by a Cord
+    if(auto body = aws_request->GetContentBody(); body) {
+      const size_t bufferSize = 1024*1024;
+      std::vector<char> buffer(bufferSize);
+      std::streampos original = body->tellg();
+      while (body->read(buffer.data(), buffer.size()) || body->gcount() > 0) {
+        payload.Append(absl::Cord(absl::string_view(buffer.data(), body->gcount())));
+      }
+      // Reset stream
+      body->clear();
+      body->seekg(original);
+    }
+
+    auto aws_headers = aws_request->GetHeaders();
+    auto headers = std::vector<std::string>{};
+    for(auto &[name, value]: aws_headers) {
+      headers.emplace_back(absl::StrCat(name, ": ", value));
+    }
+    std::string user_agent;
+    if(auto it = aws_headers.find(kUserAgentHeader); it != aws_headers.end()) {
+      user_agent = it->second;
+    }
+
+    return RequestAndPayload{
+      HttpRequest{
+        GetNameForHttpMethod(aws_request->GetMethod()),
+        aws_request->GetURIString(true),
+        std::move(user_agent),
+        std::move(headers)},
+      std::move(payload)
+    };
+  }
+
+  // Converts a tensorstore response to an Aws StandardHttpResponse
+  std::shared_ptr<AwsStandardHttpResponse> ToAwsResponse(
+      const HttpResponse & ts_response,
+      const std::shared_ptr<AwsHttpRequest> & aws_request) const {
+
+    auto aws_response = Aws::MakeShared<AwsStandardHttpResponse>(kAwsTag, aws_request);
+    aws_response->SetResponseCode(static_cast<Aws::Http::HttpResponseCode>(ts_response.status_code));
+    for(auto &[name, value]: aws_response->GetHeaders()) {
+      aws_response->AddHeader(name, value);
+    }
+    // Copy cord onto the body stream
+    // TODO: This should be avoided by subclassing Aws::IOStream
+    // to encapsulate a Cord
+    if(!ts_response.payload.empty()) {
+      aws_response->GetResponseBody() << ts_response.payload;
+    }
+
+    return aws_response;
+  }
+
+  /// Overrides the SDK mechanism for issuing AWS HttpRequests
+  /// Converts AWS HttpRequests to their tensorstore requivalent,
+  /// which is issued on the default tensorstore transport.
+  /// The tensorstore response is converted into an AWS HttpResponse.
+  std::shared_ptr<AwsHttpResponse> MakeRequest(
+    const std::shared_ptr<AwsHttpRequest> & request,
+    AwsRateLimiterInterface* readLimiter = nullptr,
+    AwsRateLimiterInterface* writeLimiter = nullptr) const override {
+      // Issue the wrapped HttpRequest on a tensorstore executor
+      auto transport = ::tensorstore::internal_http::GetDefaultHttpTransport();
+      auto [ts_request, payload] = FromAwsRequest(request);
+      ABSL_LOG(INFO) << ts_request << " " << payload;
+      auto future = transport->IssueRequest(ts_request, IssueRequestOptions(payload));
+      // TODO: if possible use a continuation (future.ExecuteWhenReady) here
+      auto response = future.value();
```

Comment:
Generally speaking, it would seem desirable to yield to the calling tensorstore Executor here, but I'm not sure that this is really possible.